data/mods/ssb/abilities.ts

```diff
@@ -196,887 +404,1759 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
-		shortDesc: "The power of Fire-type and Ground-type attacks is halved against this Pokemon.",
-		name: "Rise from the Gases",
+	bipolar: {
+		desc: "When this Pokemon switches in, it changes to two random types and gets corresponding STAB attacks.",
+		shortDesc: "This Pokemon has 2 random types and STAB moves on switch-in.",
+		name: "Bipolar",
+		isPermanent: true,
+		onSwitchIn(pokemon) {
+			const typeMap: {[key: string]: string} = {
+				Normal: "Return",
+				Fighting: "Sacred Sword",
+				Flying: "Drill Peck",
+				Poison: "Poison Jab",
+				Ground: "Earthquake",
+				Rock: "Stone Edge",
+				Bug: "Lunge",
+				Ghost: "Shadow Bone",
+				Steel: "Iron Head",
+				Electric: "Zing Zap",
+				Psychic: "Psychic Fangs",
+				Ice: "Icicle Crash",
+				Dragon: "Dual Chop",
+				Dark: "Jaw Lock",
+				Fairy: "Play Rough",
+			};
+			const types = Object.keys(typeMap);
+			this.prng.shuffle(types);
+			const newTypes = [types[0], types[1]];
+			this.add('-start', pokemon, 'typechange', newTypes.join('/'));
+			pokemon.setType(newTypes);
+			let move = this.dex.getMove(typeMap[newTypes[0]]);
+			pokemon.moveSlots[3] = pokemon.moveSlots[1];
+			pokemon.moveSlots[1] = {
+				move: move.name,
+				id: move.id,
+				pp: move.pp,
+				maxpp: move.pp,
+				target: move.target,
+				disabled: false,
+				used: false,
+				virtual: true,
+			};
+			move = this.dex.getMove(typeMap[newTypes[1]]);
+			pokemon.moveSlots[2] = {
+				move: move.name,
+				id: move.id,
+				pp: move.pp,
+				maxpp: move.pp,
+				target: move.target,
+				disabled: false,
+				used: false,
+				virtual: true,
+			};
+		},
 		isNonstandard: "Custom",
-		onSourceBasePowerPriority: 7,
-		onSourceBasePower(basePower, attacker, defender, move) {
-			if (move.type === 'Fire') {
-				return this.chainModify(0.5);
-			}
-			if (move.type === 'Ground') {
-				return this.chainModify(0.5);
+		gen: 8,
+	},
+
+	// Finland
+	windingsong: {
+		desc: "If this Pokemon's species is Alcremie, it alternates one of its moves between two different options at the end of each turn, depending on the forme of Alcremie.",
+		shortDesc: "Alcremie: alternates between moves each turn.",
+		name: "Winding Song",
+		isPermanent: true,
+		onResidual(pokemon) {
+			if (pokemon.species.baseSpecies !== 'Alcremie') return;
+			let coolMoves = [];
+			if (pokemon.species.forme === 'Lemon-Cream') {
+				coolMoves = ['Reflect', 'Light Screen'];
+			} else if (pokemon.species.forme === 'Ruby-Swirl') {
+				coolMoves = ['Refresh', 'Destiny Bond'];
+			} else if (pokemon.species.forme === 'Mint-Cream') {
+				coolMoves = ['Light of Ruin', 'Sparkling Aria'];
+			} else {
+				coolMoves = ['Infestation', 'Whirlwind'];
+			}
+			let oldMove;
+			let move;
+			if (pokemon.moves.includes(this.toID(coolMoves[0]))) {
+				oldMove = this.toID(coolMoves[0]);
+				move = this.dex.getMove(coolMoves[1]);
+			}	else if (pokemon.moves.includes(this.toID(coolMoves[1]))) {
```

Comment:
```suggestion
			} else if (pokemon.moves.includes(this.toID(coolMoves[1]))) {
```

Reply:
I'd also use some intermediate variables for `this.toID(coolMoves[1])` and so on.